### PR TITLE
Update the API documentation for Kitura

### PR DESCRIPTION
**Pull request by shmuelk**
Add class level and function level documentation to the various public APIs of Kitura.

## Motivation and Context
Have reasonable documentation.

## How Has This Been Tested?
Ran swift test.

## Checklist:
- [ ] I have submitted a [CLA form](https://github.com/IBM-Swift/CLA)
- [ ] If applicable, I have updated the documentation accordingly.
- [ ] If applicable, I have added tests to cover my changes.

### CODE REVIEW

File: Sources/Kitura/bodyParser/Part.swift [hunk 1]
```diff
@@ -16,13 +16,34 @@
 
 import Foundation
 
+// MARK Part
+
+/// A part of multi-part parsed body
```
Comment (rob-deans) [inline Sources/Kitura/bodyParser/Part.swift]: Suggestion: A part of _a_ multi-part parsed body

File: Sources/Kitura/Kitura.swift [hunk 2]
```diff
@@ -93,6 +93,8 @@ public class Kitura {
     /// Stop all registered servers
     ///
     /// Make all registered servers stop listening on their port.
+    ///
+    /// - note: All of the registered servers are unregistered afr they are stopped.
```
Comment (rob-deans) [inline Sources/Kitura/Kitura.swift]: Typo? afr -> after

File: Sources/Kitura/bodyParser/ParsedBody.swift [hunk 3]
```diff
@@ -19,23 +19,29 @@ import Foundation
 
 // MARK ParsedBody
 
-/// The result of body parsing
+/// The result of parsing the body of the request.
 ///
-/// - SeeAlso: `BodyParser.parse(_ mesage:, contentType:)->ParsedBody?`
+/// When a body of a request is parsed the results of the parsing are placed 
+/// in case of this enum based on Content-Type.
```
Comment (rob-deans) [inline Sources/Kitura/bodyParser/ParsedBody.swift]: Suggestion: in _the associated value of the enum case_ based on Content-Type

File: Sources/Kitura/LinkParameter.swift [hunk 4]
```diff
@@ -14,13 +14,32 @@
  * limitations under the License.
  */
 
+// MARK LinkParameter
+
+/// Possible parameters of Link HTTP header.
```
Comment (rob-deans) [inline Sources/Kitura/LinkParameter.swift]: Suggestion: Possible parameters of _a_ Link HTTP header.

File: Sources/Kitura/Headers.swift [hunk 5]
```diff
@@ -17,41 +17,51 @@
...
         return headers.endIndex
     }
     
+    /// The type of an Index of the `Headers` collection.
     public typealias HeadersIndex = DictionaryIndex<String, [String]>
     
+    
+    /// Get the value of a HTTP header
+    ///
+    /// - Parameter key: The HTTP header key who's value is to be gotten
```
Comment (rob-deans) [inline Sources/Kitura/Headers.swift]: Suggestion: gotten -> retrieved

File: Sources/Kitura/Router.swift [hunk 6]
```diff
@@ -21,7 +21,14 @@ import KituraTemplateEngine
 
 // MARK Router
 
+/// The `Router` class provides the external intreface for the routing of requestes to
```
Comment (rob-deans) [inline Sources/Kitura/Router.swift]: Typo: intreface -> interface, requestes -> requests

File: Sources/Kitura/Router.swift [hunk 7]
```diff
@@ -72,10 +85,17 @@ public class Router {
         defaultEngineFileExtension = nil
     }
 
+    /// Register a templating engine. The templating engine will handle files in the `viewsPath`
+    /// that match the extension it supports.
+    ///
+    /// - Parameter templateEngine: The templating engineto register.
```
Comment (rob-deans) [inline Sources/Kitura/Router.swift]: Typo: engineto -> engine to

File: Sources/Kitura/Router.swift [hunk 8]
```diff
@@ -72,10 +85,17 @@ public class Router {
...
     public func add(templateEngine: TemplateEngine) {
         templateEngines[templateEngine.fileExtension] = templateEngine
     }
 
+    /// Render a template using a context
+    ///
+    /// - Parameter template:
```
Comment (rob-deans) [inline Sources/Kitura/Router.swift]: Missing description?

File: Sources/Kitura/Router.swift [hunk 9]
```diff
@@ -172,9 +207,12 @@ extension Router : ServerDelegate {
...
     ///
-    /// - Parameter request: the server request
-    /// - Parameter response: the server response
-    /// - Parameter callback:
+    /// - Parameter request: The `RouterRequest` object that is used to work with
+    ///                     the incoming request.
+    /// - Parameter response: The `RouterResponse` object used to send responses
+    ///                      to the HTTP request.
+    /// - Parameter callback: The closure to invoke to cause the router to inspect the
+    ///                  path in th elist of paths.
```
Comment (rob-deans) [inline Sources/Kitura/Router.swift]: Typo: the elist -> the list

File: Sources/Kitura/RouterMethod.swift [hunk 10]
```diff
@@ -14,44 +14,110 @@
  * limitations under the License.
  */
 
-/// Values for Router methods (Get, Post, Put, Delete, etc)
+/// An enum to describe the HTTP method (Get, Post, Put, Delete, etc) of an HTTP
+/// request. In general they match the actual HTTP methods by th esame name. There
```
Comment (rob-deans) [inline Sources/Kitura/RouterMethod.swift]: Typo: th esame -> the same

File: Sources/Kitura/RouterMethod.swift [hunk 11]
```diff
@@ -14,44 +14,110 @@
...
+/// are two special ones, used by `Router` when building up the set of mapiings
```
Comment (rob-deans) [inline Sources/Kitura/RouterMethod.swift]: Typo: mapiings -> mappings

File: Sources/Kitura/RouterMethod.swift [hunk 12]
```diff
@@ -14,44 +14,110 @@
...
+/// betwem paths and handlers or middle ware. They are:
```
Comment (rob-deans) [inline Sources/Kitura/RouterMethod.swift]: Type: betwem -> between, middle ware -> middleware ?

File: Sources/Kitura/RouterMethod.swift [hunk 13]
```diff
@@ -14,44 +14,110 @@
...
     case connect = "CONNECT"
+    
+    /// Used to mark an error handler in the list of router mappings.
     case error = "ERROR"
+    
+    /// Created when creating instances of this enum from a string that doesn't match any of the other
+    /// values.
     case unknown = "UNKNOWN"
 
+    /// Convert a string to a `RouterMEthod` instance.
```
Comment (rob-deans) [inline Sources/Kitura/RouterMethod.swift]: Typo: 'RouterMEthod' -> 'RouterMethod'

File: Sources/Kitura/RouterMiddleware.swift [hunk 14]
```diff
@@ -16,12 +16,23 @@
...
+/// in nature and not tied to a specific request.
 public protocol RouterMiddleware {
 
-    /// The handle operation
+    /// Handle an incoming HTTP request.
     ///
-    /// - Parameter request: the router request
-    /// - Parameter response: the router response
-    /// - Parameter next: the closure to the next operation
+    /// - Parameter request: The `RouterRequest` object used to get inormation
```
Comment (rob-deans) [inline Sources/Kitura/RouterMiddleware.swift]: Typo: inormation -> information

File: Sources/Kitura/Router.swift [hunk 15]
```diff
@@ -104,22 +124,35 @@ public class Router {
         let filePath =  viewsPath + resourceWithExtension
         return try templateEngine.render(filePath: filePath, context: context)
     }
-
+    
+    // MARK: Sub router
+    
+    /// Setup a "sub router" to handle requests. This can make it easier to
+    /// build a server that server a large set of paths, by breaking it up
```
Comment (rob-deans) [inline Sources/Kitura/Router.swift]: Typo: that _server_ -> that _serves_

File: Sources/Kitura/Headers.swift [hunk 16]
```diff
@@ -66,24 +76,35 @@ extension Headers: Collection {
...
     public subscript(position: HeadersIndex) -> (String, String?) {
         get {
             let (key, value) = headers[position]
             return (key, value.first)
         }
     }
     
+    /// Get the next Index in the `Headers` collection after the one specified.
+    ///
+    /// - Parameter after: The Index before the Index to be returned.
```
Comment (rob-deans) [inline Sources/Kitura/Headers.swift]: Suggestion: before -> after

File: Sources/Kitura/RouterMiddlewareGenerator.swift [hunk 17]
```diff
@@ -24,11 +24,11 @@ public class RouterMiddlewareGenerator: RouterMiddleware {
 
     /// Initialize a RouterMiddlewareGenerator
     ///
-    /// - Parameter handler: the closure to be called to handle requests
+    /// - Parameter handler: The closure that is of the type `RouterHAndler` to be
```
Comment (rob-deans) [inline Sources/Kitura/RouterMiddlewareGenerator.swift]: Typo: 'RouterHAndler' -> 'RouterHandler'

File: Sources/Kitura/RouterResponse.swift [hunk 18]
```diff
@@ -338,8 +345,8 @@ public class RouterResponse {
     }
 }
 
-/// Type alias for "Before flush" (i.e. before headers and body are written) lifecycle handler
+/// Type alias for "Before flush" (i.e. before headers and body are written) lifecycle handler..
```
Comment (rob-deans) [inline Sources/Kitura/RouterResponse.swift]: Typo: .. -> .

File: Sources/Kitura/staticFileServer/StaticFileServer.swift [hunk 19]
```diff
@@ -17,15 +17,24 @@
 import Foundation
 
 // MARK: StaticFileServer
-
-///
+.
+/// A router middleware that serves static files from a given path
```
Comment (rob-deans) [inline Sources/Kitura/staticFileServer/StaticFileServer.swift]: Missing period at end, seems to be on the line before

File: Scripts/generate_router_verbs.sh [hunk 20]
```diff
@@ -43,24 +43,68 @@ EOF
...
+  DOC_TEXT_2=""
+else
+  DOC_TEXT_1="HTTP $VERB_UPPER_CASE"
+  DOC_TEXT_2="s"
+fi
 cat <<EOF >> ${OUTPUT_FILE}
     // MARK: $VERB
 
+    /// Setup a set of one or more closures of the type \`RouterHandler\` that will be
+    /// invoked when $DOC_TEXT_1 request$DOC_TEXT_2 comes to the server. If a path pattern is
```
Comment (rob-deans) [inline Scripts/generate_router_verbs.sh]: Typo?: request$DOC... -> ?

File: Scripts/generate_router_verbs.sh [hunk 21]
```diff
@@ -43,24 +43,68 @@ EOF
...
+    /// specified, the handlers will only be invoked if the pattern is matched.
+    ///
+    /// - Parameter path: An optional String specifying the patern that needs to be
```
Comment (rob-deans) [inline Scripts/generate_router_verbs.sh]: Typo: patern -> pattern

File: Scripts/generate_router_verbs.sh [hunk 22]
```diff
@@ -43,24 +43,68 @@ EOF
...
     @discardableResult
     public func $VERB_LOW_CASE(_ path: String?=nil, handler: [RouterHandler]) -> Router {
         return routingHelper(.$VERB_LOW_CASE, pattern: path, handler: handler)
     }
 
+    /// Setup a set of one or more \`RouterMiddleware\` that will be
+    /// invoked when $DOC_TEXT_1 request$DOC_TEXT_2 comes to the server. If a path pattern is
+    /// specified, the \`RouterMiddleware\` will only be invoked if the pattern is matched.
+    ///
+    /// - Parameter path: An optional String specifying the patern that needs to be
```
Comment (rob-deans) [inline Scripts/generate_router_verbs.sh]: Type: patern -> pattern

File: Scripts/generate_router_verbs.sh [hunk 23]
```diff
@@ -43,24 +43,68 @@ EOF
...
     @discardableResult
     public func $VERB_LOW_CASE(_ path: String?=nil, allowPartialMatch: Bool = true, middleware: RouterMiddleware...) -> Router {
         return routingHelper(.$VERB_LOW_CASE, pattern: path, allowPartialMatch: allowPartialMatch, middleware: middleware)
     }
 
+    /// Setup an array of one or more \`RouterMiddleware\` that will be
+    /// invoked when $DOC_TEXT_1 request$DOC_TEXT_2 comes to the server. If a path pattern is
+    /// specified, the \`RouterMiddleware\` will only be invoked if the pattern is matched.
+    ///
+    /// - Parameter path: An optional String specifying the patern that needs to be
```
Comment (rob-deans) [inline Scripts/generate_router_verbs.sh]: Typo: patern -> pattern

File: Sources/Kitura/RouterHTTPVerbs_generated.swift [hunk 24]
```diff
@@ -19,567 +19,1539 @@
 extension Router {
     // MARK: All
 
+    /// Setup a set of one or more closures of the type `RouterHandler` that will be
+    /// invoked when any request comes to the server. If a path pattern is
+    /// specified, the handlers will only be invoked if the pattern is matched.
+    ///
+    /// - Parameter path: An optional String specifying the patern that needs to be
```
Comment (rob-deans) [inline Sources/Kitura/RouterHTTPVerbs_generated.swift]: Typo: patern -> pattern - and in the resulting places below where it has also been pasted

**Comment by rob-deans** [comment]
Noted that there were some inconsistencies when writing about initializing class e.g:
`Initialize a `Headers``
`Initialize a RouterResponse instance`
In the docs

File: Sources/Kitura/bodyParser/Part.swift (hunk 1 above)
Comment (shmuelk) [inline Sources/Kitura/bodyParser/Part.swift]: The current text is: 

    A part of a parsed multi-part form body.

File: Sources/Kitura/Kitura.swift (hunk 2 above)
Comment (shmuelk) [inline Sources/Kitura/Kitura.swift]: Done

File: Sources/Kitura/bodyParser/ParsedBody.swift (hunk 3 above)
Comment (shmuelk) [inline Sources/Kitura/bodyParser/ParsedBody.swift]: Done

File: Sources/Kitura/Headers.swift (hunk 5 above)
Comment (shmuelk) [inline Sources/Kitura/Headers.swift]: Done

File: Sources/Kitura/Router.swift (hunk 6 above)
Comment (shmuelk) [inline Sources/Kitura/Router.swift]: Done

File: Sources/Kitura/Router.swift (hunk 7 above)
Comment (shmuelk) [inline Sources/Kitura/Router.swift]: Done.

File: Sources/Kitura/Router.swift (hunk 8 above)
Comment (shmuelk) [inline Sources/Kitura/Router.swift]: text is now:    
    /// Render a template using a context
    ///
    /// - Parameter template: The path to the template file to be rendered.
    /// - Parameter context: A Dictionary of variables to be used by the
    ///                     template engine while rendering the template.
    ///
    /// - Returns: The content generated by rendering the template.
    /// - Throws: Any error thrown by the Templating Engine when it fails to
    ///          render the template.

File: Sources/Kitura/Router.swift (hunk 9 above)
Comment (shmuelk) [inline Sources/Kitura/Router.swift]: Done.

File: Sources/Kitura/RouterMethod.swift (hunk 10 above)
Comment (shmuelk) [inline Sources/Kitura/RouterMethod.swift]: Done

File: Sources/Kitura/RouterMethod.swift (hunk 11 above)
Comment (shmuelk) [inline Sources/Kitura/RouterMethod.swift]: Done.

File: Sources/Kitura/RouterMethod.swift (hunk 12 above)
Comment (shmuelk) [inline Sources/Kitura/RouterMethod.swift]: Done.

File: Sources/Kitura/RouterMethod.swift (hunk 13 above)
Comment (shmuelk) [inline Sources/Kitura/RouterMethod.swift]: Done

File: Sources/Kitura/RouterMiddleware.swift (hunk 14 above)
Comment (shmuelk) [inline Sources/Kitura/RouterMiddleware.swift]: Done

File: Sources/Kitura/Router.swift (hunk 15 above)
Comment (shmuelk) [inline Sources/Kitura/Router.swift]: Done.

File: Sources/Kitura/Headers.swift (hunk 16 above)
Comment (shmuelk) [inline Sources/Kitura/Headers.swift]: Instead changed line 94 to read:

Parameter after: The Index who's successor is to be returned.

File: Sources/Kitura/RouterMiddlewareGenerator.swift (hunk 17 above)
Comment (shmuelk) [inline Sources/Kitura/RouterMiddlewareGenerator.swift]: Done

File: Sources/Kitura/RouterResponse.swift (hunk 18 above)
Comment (shmuelk) [inline Sources/Kitura/RouterResponse.swift]: Done

File: Sources/Kitura/staticFileServer/StaticFileServer.swift (hunk 19 above)
Comment (shmuelk) [inline Sources/Kitura/staticFileServer/StaticFileServer.swift]: Done

File: Scripts/generate_router_verbs.sh (hunk 21 above)
Comment (shmuelk) [inline Scripts/generate_router_verbs.sh]: Done

File: Scripts/generate_router_verbs.sh (hunk 22 above)
Comment (shmuelk) [inline Scripts/generate_router_verbs.sh]: Done

File: Scripts/generate_router_verbs.sh (hunk 23 above)
Comment (shmuelk) [inline Scripts/generate_router_verbs.sh]: Done

File: Sources/Kitura/RouterHTTPVerbs_generated.swift (hunk 24 above)
Comment (shmuelk) [inline Sources/Kitura/RouterHTTPVerbs_generated.swift]: Done

**Comment by shmuelk** [comment]
I have tried to make the constructors more consistent.

File: Sources/Kitura/Headers.swift (hunk 16 above)
Comment (ianpartridge) [inline Sources/Kitura/Headers.swift]: `whose` not `who's`.

**Pull request merged**